### PR TITLE
Multi Ref Support

### PR DESCRIPTION
**Pull request by ammokhov**
*Issue #, if available:*

*Description of changes:*

## Valid Schema Case:

adding multi `$ref` support to enrich schema capabilities to support multiple types; 
usecases:
```json
{
    "properties": {
        "propertyA": {
            "oneOf": [
                {"$ref": "#..."},
                {"$ref": "#..."},
                {"type": "..."}
            ]
        }
    }
}
```
--------------------------------------
### Resource Schema:

```json
{
    "typeName": "AWS::Valid::TypeName",
    "description": "a test schema",
    "definitions": {
        "ComplexTypeWithOnePrimitive": {
            "type": "object",
            "properties": {
                "PrimitiveProperty": {
                    "type": "string"
                }
            }
        },
        "ComplexTypeWithMultiplePrimitives": {
            "type": "object",
            "properties": {
                "PrimitiveProperty1": {
                    "type": "string"
                },
                "PrimitiveProperty2": {
                    "type": "string"
                }
            }
        }
    },
    "properties": {
        "primaryID": {
            "type": "string"
        },
        "PropertyWithTwoComplexTypes": {
            "oneOf": [
                {
                    "$ref": "#/definitions/ComplexTypeWithOnePrimitive"
                },
                {
                    "$ref": "#/definitions/ComplexTypeWithMultiplePrimitives"
                }
            ]
        }
    },
    "additionalProperties": false,
    "createOnlyProperties": [
        "/properties/primaryID"
    ],
    "primaryIdentifier": [
        "/properties/primaryID"
    ]
}
```

#### README.md

```md
# AWS::Valid::TypeName

a test schema

## Syntax

To declare this entity in your AWS CloudFormation template, use the following syntax:

### JSON

<pre>
{
    "Type" : "AWS::Valid::TypeName",
    "Properties" : {
        "<a href="#primaryid" title="primaryID">primaryID</a>" : <i>String</i>,
        "<a href="#propertywithtwocomplextypes" title="PropertyWithTwoComplexTypes">PropertyWithTwoComplexTypes</a>" : <i><a href="complextypewithoneprimitive.md">ComplexTypeWithOnePrimitive</a>, <a href="complextypewithmultipleprimitives.md">ComplexTypeWithMultiplePrimitives</a></i>
    }
}
</pre>

### YAML

<pre>
Type: AWS::Valid::TypeName
Properties:
    <a href="#primaryid" title="primaryID">primaryID</a>: <i>String</i>
    <a href="#propertywithtwocomplextypes" title="PropertyWithTwoComplexTypes">PropertyWithTwoComplexTypes</a>: <i><a href="complextypewithoneprimitive.md">ComplexTypeWithOnePrimitive</a>, <a href="complextypewithmultipleprimitives.md">ComplexTypeWithMultiplePrimitives</a></i>
</pre>

## Properties

#### primaryID

_Required_: No

_Type_: String

_Update requires_: [Replacement](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-replacement)

#### PropertyWithTwoComplexTypes

_Required_: No

_Type_: <a href="complextypewithoneprimitive.md">ComplexTypeWithOnePrimitive</a>, <a href="complextypewithmultipleprimitives.md">ComplexTypeWithMultiplePrimitives</a>

_Update requires_: [No interruption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-no-interrupt)

## Return Values

### Ref

When you pass the logical ID of this resource to the intrinsic `Ref` function, Ref returns the primaryID.

```

#### complextypewithoneprimitive.md

```md
# AWS::Valid::TypeName ComplexTypeWithOnePrimitive

## Syntax

To declare this entity in your AWS CloudFormation template, use the following syntax:

### JSON

<pre>
{
    "<a href="#primitiveproperty" title="PrimitiveProperty">PrimitiveProperty</a>" : <i>String</i>
}
</pre>

### YAML

<pre>
<a href="#primitiveproperty" title="PrimitiveProperty">PrimitiveProperty</a>: <i>String</i>
</pre>

## Properties

#### PrimitiveProperty

_Required_: No

_Type_: String

_Update requires_: [No interruption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-no-interrupt)

```

#### complextypewithmultipleprimitives.md

```md
# AWS::Valid::TypeName ComplexTypeWithMultiplePrimitives

## Syntax

To declare this entity in your AWS CloudFormation template, use the following syntax:

### JSON

<pre>
{
    "<a href="#primitiveproperty1" title="PrimitiveProperty1">PrimitiveProperty1</a>" : <i>String</i>,
    "<a href="#primitiveproperty2" title="PrimitiveProperty2">PrimitiveProperty2</a>" : <i>String</i>
}
</pre>

### YAML

<pre>
<a href="#primitiveproperty1" title="PrimitiveProperty1">PrimitiveProperty1</a>: <i>String</i>
<a href="#primitiveproperty2" title="PrimitiveProperty2">PrimitiveProperty2</a>: <i>String</i>
</pre>

## Properties

#### PrimitiveProperty1

_Required_: No

_Type_: String

_Update requires_: [No interruption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-no-interrupt)

#### PrimitiveProperty2

_Required_: No

_Type_: String

_Update requires_: [No interruption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-no-interrupt)

```

#### ResourceModel.java

```java
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class ResourceModel {
    @JsonIgnore
    public static final String TYPE_NAME = "AWS::Valid::TypeName";

    @JsonIgnore
    public static final String IDENTIFIER_KEY_PRIMARYID = "/properties/primaryID";

    @JsonProperty("primaryID")
    private String primaryID;

    @JsonProperty("PropertyWithTwoComplexTypes")
    private Object propertyWithTwoComplexTypes;

    @JsonIgnore
    public JSONObject getPrimaryIdentifier() {
        final JSONObject identifier = new JSONObject();
        if (this.getPrimaryID() != null) {
            identifier.put(IDENTIFIER_KEY_PRIMARYID, this.getPrimaryID());
        }

        // only return the identifier if it can be used, i.e. if all components are present
        return identifier.length() == 1 ? identifier : null;
    }

    @JsonIgnore
    public List<JSONObject> getAdditionalIdentifiers() {
        final List<JSONObject> identifiers = new ArrayList<JSONObject>();
        // only return the identifiers if any can be used
        return identifiers.isEmpty() ? null : identifiers;
    }
}
```

#### ComplexTypeWithMultiplePrimitives.java
```java
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class ComplexTypeWithMultiplePrimitives {
    @JsonProperty("PrimitiveProperty1")
    private String primitiveProperty1;

    @JsonProperty("PrimitiveProperty2")
    private String primitiveProperty2;

}
```

#### ComplexTypeWithOnePrimitive.java

```java
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class ComplexTypeWithOnePrimitive {
    @JsonProperty("PrimitiveProperty")
    private String primitiveProperty;

}
```
--------------------------------------------

### Rationale:
schema flattener combines multiple types into a collection of types: `{ "oneOf": [{"type": "string"},{"type": "integer"}] }` => `"type": ["string", "integer"]`
similarly schema flattener combines `$ref` types in a similar way `{ "oneOf": [{"$ref": "..."},{"$ref": "..."}] }` => `"type": [(..., ...), (..., ....)]`

having know this flattened types, schema resolver will resolve list as `MULTIPLE` hence both will be defaulted to `java.lang.Object`/`Any`

------------------------------------------

#### Type and oneOf on the same level
Such schema is considered to be valid by json engine but any input would be invalid as two attributes are mutually exclusive:
```json
{
    "property": {
        "type": "string",
        "oneOf": [
            { "type": "boolean" },
            { "type": "integer" }
        ]
    }
}
```
CLI will prioritize `type` over combiner:
resulted Resource Model
```java
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class ResourceModel {
    ...
    @JsonProperty("property")
    private String property;
```

#### Ref and oneOf on the same level
Such schema is considered to be valid by json engine but any input would be invalid as two attributes are mutually exclusive:
```json
{
    "property": {
        "$ref": "#/definitions/ComplexType",
        "oneOf": [
            { "type": "boolean" },
            { "type": "integer" }
        ]
    }
}
```
CLI will prioritize `$ref` over combiner:
resulted Resource Model
```java
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class ResourceModel {
    ...
    @JsonProperty("property")
    private ComplexType property;
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by aygold92 — COMMENTED** [review]
I don't see any proof that this works end to end, or what different use cases look like.  How was this tested?

File: src/rpdk/core/jsonutils/flattener.py [hunk 1]
```diff
@@ -158,18 +157,25 @@ def _flatten_combiners(self, sub_schema, path):
...
-                walked_schema = self._walk(nested_schema, ref_path)
-
-                # if no other schema is referencing the ref_path,
-                # we no longer need the refkey since the properties will be squashed
-                if ref_path_is_used:
-                    resolved_schema = self._schema_map.get(ref_path)
-                else:
-                    resolved_schema = self._schema_map.pop(ref_path, walked_schema)
-                schema_merge(sub_schema, resolved_schema, path)
+            else:
```
[the remark this answers is not in the document]
Comment (ammokhov) [reply src/rpdk/core/jsonutils/flattener.py:160]: yep, we can do `pass` instead to be consistent with the style that is used in cli

File: src/rpdk/core/jsonutils/flattener.py [hunk 2]
```diff
@@ -158,18 +157,25 @@ def _flatten_combiners(self, sub_schema, path):
...
+                    if ref_path_is_used:
+                        resolved_schema = self._schema_map.get(ref_path)
+                    else:
+                        resolved_schema = self._schema_map.pop(ref_path, walked_schema)
+                    schema_merge(sub_schema, resolved_schema, path)
+
+        if REF in sub_schema and isinstance(
+            sub_schema.get(TYPE), list
+        ):  # check if there are conflicting $ref and type at the same sub schema
+            sub_schema.pop(REF)
```
[the remark this answers is not in the document]
Comment (ammokhov) [reply src/rpdk/core/jsonutils/flattener.py:178]: you mean
```json
{
    "type": "object",
    "$ref": "#/some/ref",
    "allOf": [
        {"type": "boolean"},
        {"type": "integer"},
        {"$ref": "#/..."}
    ]
}
```

File: tests/data/schema/target_output/multiref.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+_Type_: <a href="complextypewithoneprimitive.md">ComplexTypeWithOnePrimitive</a>, <a href="complextypewithmultipleprimitives.md">ComplexTypeWithMultiplePrimitives</a>, <a href="complextypewithcircularref.md">ComplexTypeWithCircularRef</a>
+
+_Update requires_: [No interruption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-no-interrupt)
+
+#### PropertyWithMultipleComplexTypesAndOnePrimitive
+
+_Required_: No
+
+_Type_: <a href="complextypewithoneprimitive.md">ComplexTypeWithOnePrimitive</a>, <a href="complextypewithmultipleprimitives.md">ComplexTypeWithMultiplePrimitives</a>, <a href="complextypewithcircularref.md">ComplexTypeWithCircularRef</a>, Map
```
Comment (aygold92) [inline tests/data/schema/target_output/multiref.md:90]: brought this up before, but we're generating a link to `ComplexTypeWithOnePrimitive` here, are we actually generating these type objects in docs or in codegen? I don't see it in this file

**Review by aygold92 — APPROVED** [review]

File: src/rpdk/core/jsonutils/utils.py [hunk 4]
```diff
@@ -138,11 +140,17 @@ def schema_merge(target, src, path):  # noqa: C901
             try:
                 target[key] = schema_merge(target_schema, src_schema, next_path)
             except TypeError:
-                if key == "type":
```
Comment (jotompki) [inline src/rpdk/core/jsonutils/utils.py:141]: want to put this up above but can't on github, but can we add the cases to test this new logic to the doctests? would like it just to visualize what the additions to the method are doing w minimal examples.

File: src/rpdk/core/jsonutils/utils.py [hunk 5]
```diff
@@ -138,11 +140,17 @@ def schema_merge(target, src, path):  # noqa: C901
...
-                    if isinstance(target_schema, list):
-                        target_schema.append(src_schema)
-                        continue
-                    target[key] = [target_schema, src_schema]
+                if key in (TYPE, REF):  # combining multiple $ref and types
```
Comment (jotompki) [inline src/rpdk/core/jsonutils/utils.py:143]: what happens if key is ref here? handled the same as multiple types?

File: src/rpdk/core/jsonutils/utils.py [hunk 6]
```diff
@@ -138,11 +140,17 @@ def schema_merge(target, src, path):  # noqa: C901
...
+                    if isinstance(
+                        target.get(TYPE), list
+                    ):  # check not to add duplicates
+
+                        if src_schema not in target[TYPE]:
+                            target[TYPE].append(src_schema)
```
Comment (jotompki) [inline src/rpdk/core/jsonutils/utils.py:149]: can we use set union or something here to cut down on if statements and get rid of that pylint disable

File: src/rpdk/core/jsonutils/utils.py [hunk 7]
```diff
@@ -138,11 +140,17 @@ def schema_merge(target, src, path):  # noqa: C901
...
+                    else:
+                        target[TYPE] = [target_schema]
+                        if src_schema != target_schema:
```
Comment (jotompki) [inline src/rpdk/core/jsonutils/utils.py:152]: same here

File: src/rpdk/core/jsonutils/utils.py (hunk 5 above)
Comment (ammokhov) [reply src/rpdk/core/jsonutils/utils.py:143]:
```
>>> a, b = {'Foo': {'$ref': 'a'}}, {'Foo': {'$ref': 'b'}}
>>> schema_merge(a, b, ('foo',))
{'Foo': { 'type': [(..., ...), (..., ...)]}}
```

**Review by sungkkim — APPROVED** [review]

**Review by jotompki — APPROVED** [review]

**Pull request merged**